### PR TITLE
CI: run pytype and ruff on Python code

### PR DESCRIPTION
**Pull request by dfandrich**
There are a few TODOs, but only one of them needs to be addressed now as it
will currently cause a check failure.

### CODE REVIEW

File: .github/workflows/checksrc.yml [hunk 1]
```diff
@@ -41,23 +41,35 @@ jobs:
       - name: check
         run: git ls-files -z "*.[ch]" | xargs -0 -n1 ./scripts/checksrc.pl
 
-  codespell-cmakelint:
-    runs-on: ubuntu-latest
+  codespell-cmake-python:
```
Comment (bagder) [inline .github/workflows/checksrc.yml:44]: To me, using `cmakelint` in the name seems to tell us more relevant info about the job than `cmake-python` does...

**Comment by dfandrich** [comment]
I just wanted to shorten the name a bit. I could just change it to codespell-cmakelint-pytype-ruff

**Review by icing — APPROVED** [review]

**Comment by dfandrich** [comment]
@icing what's the right solution to that [pytest failure](https://github.com/curl/curl/actions/runs/11061488918/job/30734202235?pr=15067) in curl.py? It looks like it's found a legitimate bug in the code.

**Comment by icing** [comment]
> @icing what's the right solution to that [pytest failure](https://github.com/curl/curl/actions/runs/11061488918/job/30734202235?pr=15067) in curl.py? It looks like it's found a legitimate bug in the code.

Hmm, strange. the `self._parse_headerfile()` should have added a response. But if this checker only simulates and empty file or otherwise short-circuits it...then I change the if to

```python
if r.son and r.response:
    r.response["json"] = r.json
```

**Comment by dfandrich** [comment]
I think the issue is that r.response is a @property method and can't be an lvalue.

**Comment by icing** [comment]
> I think the issue is that r.response is a @Property method and can't be an lvalue.

I think we can remove it. It seems a residue from migrating it from other test suite I used before.

**Comment by dfandrich** [comment]
Actually, my rationale was wrong. The issue is that the r.response could return None which can't take a subscript. I've just deleted it instead as you suggest.

**Comment by dfandrich** [comment]
I should note that there's also a dependency on Python 3.11 (or 3.12) in curl.py line 99. `statistics.fmean` doesn't take a `weights` parameter in Python 3.10 or below.

**Comment by dfandrich** [comment]
Fixing that one problem unblocked checks on more files, some of which had no obvious fixes. @icing please take a look at the two new FIXUP commits and see if they make sense. Specifically:
* I added some asserts to ensure that credentials were never None; maybe these should be turned into run-time checks or the types made non-Optional[] instead.
* In VsFTPD I replaced one `subprocess.run()` with `subprocess.Popen()` because the API is different and some later calls (e.g. to `poll()`) would have failed. The conversion wasn't perfect, though, because the combined stdout and stderr are now being passed into `ExecResult`, unlike before, but I don't know if this is an issue or not.
* I suspect the `start = datetime.now()` line just before that ExecResult is in the wrong place, because it's only going to ever measure microseconds when it's used on the next line.
* I'm not even sure this code is run, as a previous version with errors successfully passed the pytest suite.

**Comment by icing** [comment]
> Fixing that one problem unblocked checks on more files, some of which had no obvious fixes. @icing please take a look at the two new FIXUP commits and see if they make sense. Specifically:
> 
> * I added some asserts to ensure that credentials were never None; maybe these should be turned into run-time checks or the types made non-Optional[] instead.

I think your change is fine here.

> * In VsFTPD I replaced one `subprocess.run()` with `subprocess.Popen()` because the API is different and some later calls (e.g. to `poll()`) would have failed. The conversion wasn't perfect, though, because the combined stdout and stderr are now being passed into `ExecResult`, unlike before, but I don't know if this is an issue or not.

I don't see that in the diff. Can you point out the location you mean?

> * I suspect the `start = datetime.now()` line just before that ExecResult is in the wrong place, because it's only going to ever measure microseconds when it's used on the next line.

Agree. That line should be before the run. Same seems to be true for httpd.py:129

> * I'm not even sure this code is run, as a previous version with errors successfully passed the pytest suite.

**Comment by dfandrich** [comment]
> > * In VsFTPD I replaced one `subprocess.run()` with `subprocess.Popen()` because the API is different and some later calls (e.g. to `poll()`) would have failed. The conversion wasn't perfect, though, because the combined stdout and stderr are now being passed into `ExecResult`, unlike before, but I don't know if this is an issue or not.
> 
> I don't see that in the diff. Can you point out the location you mean?

[This](https://github.com/curl/curl/commit/43123717a9686bb6d9974840e4e85840253c01a9#diff-40e8c8b7b042fa5414b6c337262f619bce1215ab181c270de44cc5c25bdc8140R159) is the code I mean.

**Comment by icing** [comment]
> > > * In VsFTPD I replaced one `subprocess.run()` with `subprocess.Popen()` because the API is different and some later calls (e.g. to `poll()`) would have failed. The conversion wasn't perfect, though, because the combined stdout and stderr are now being passed into `ExecResult`, unlike before, but I don't know if this is an issue or not.
> > 
> > 
> > I don't see that in the diff. Can you point out the location you mean?
> 
> [This](https://github.com/curl/curl/commit/43123717a9686bb6d9974840e4e85840253c01a9#diff-40e8c8b7b042fa5414b6c337262f619bce1215ab181c270de44cc5c25bdc8140R159) is the code I mean.

Thanks. I do not see the `_run()` method being used in vsftpd.py. It must be a copy&paste leftover from other server handling classes. I think we should just remove it.

**Pull request closed without merging**